src/tx_display.c

```diff
@@ -0,0 +1,523 @@
+/*******************************************************************************
+*   (c) 2018 - 2023 Zondax AG
+*
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+********************************************************************************/
+#ifdef __cplusplus
+#pragma clang diagnostic push
+#pragma ide diagnostic ignored "misc-no-recursion"
+#endif
+
+#include "coin.h"
+#include "app_mode.h"
+#include "tx_display.h"
+#include "tx_parser.h"
+#include "parser_impl.h"
+#include <zxmacros.h>
+
+#define NUM_REQUIRED_ROOT_PAGES 7
+
+const char *get_required_root_item(root_item_e i) {
+    switch (i) {
+        case root_item_chain_id:
+            return "chain_id";
+        case root_item_account_number:
+            return "account_number";
+        case root_item_sequence:
+            return "sequence";
+        case root_item_memo:
+            return "memo";
+        case root_item_msgs:
+            return "msgs";
+        case root_item_source:
+            return "source";
+        case root_item_data:
+            return "data";
+        default:
+            return "?";
+    }
+}
+
+#ifdef __cplusplus
+#pragma clang diagnostic push
+#pragma ide diagnostic ignored "bugprone-branch-clone"
+#endif
+
+__Z_INLINE uint8_t get_root_max_level(root_item_e i) {
+    switch (i) {
+        case root_item_chain_id:
+            return 1;
+        case root_item_account_number:
+            return 1;
+        case root_item_sequence:
+            return 1;
+        case root_item_memo:
+            return 1;
+        case root_item_msgs:
+            return 2;
+        case root_item_source:
+            return 1;
+        case root_item_data:
+            return 1;
+        default:
+            return 0;
+    }
```

Comment:
the `case` statements could be grouped when they return the same value?

Reply:
Ok.